src/web/lib/middleware.js

```diff
@@ -0,0 +1,104 @@
+import { actions, selectors, themeChangeActions } from "../../lib/store";
+
+export default function({
+  postMessage,
+  urlEncodeTheme,
+  storage: { imageStorage }
+}) {
+  const updateExtensionThemeMiddleware = ({ getState }) => next => action => {
+    const returnValue = next(action);
+    const meta = action.meta || {};
+    if (!meta.skipAddon && themeChangeActions.includes(action.type)) {
+      postMessage("setTheme", { theme: selectors.theme(getState()) });
+    }
+    return returnValue;
+  };
+
+  const updateHistoryMiddleware = ({ getState }) => next => action => {
+    const returnValue = next(action);
+    const meta = action.meta || {};
+    if (!meta.skipHistory && themeChangeActions.includes(action.type)) {
+      const state = getState();
+      const theme = selectors.theme(state);
+      const hasCustomBackgrounds = selectors.themeHasCustomBackgrounds(state);
+      urlEncodeTheme({ theme, hasCustomBackgrounds }).then(url =>
+        window.history.pushState({ theme }, "", url)
+      );
+    }
+    return returnValue;
+  };
+
+  const imageManagementMiddleware = ({
+    getState,
+    dispatch
+  }) => next => action =>
+    action.type in imageHandlers
+      ? imageHandlers[action.type](getState, dispatch, next, action)
+      : next(action);
+
+  const imageHandlers = {
+    [actions.images.updateImage]: (getState, dispatch, next, action) => {
+      const rv = next(action);
+      const { name } = action.payload;
+      const state = getState();
+      const images = selectors.themeCustomImages(state);
+      const image = images[name];
+      if (image.importing) {
+        const { importing, ...importedImage } = image; // eslint-disable-line no-unused-vars
+        postMessage("updateImage", { image: importedImage });
+        imageStorage.put(name, importedImage);
+        dispatch(actions.images.updateImage({ name, importing: false }));
+      }
+      return rv;
+    },
+    [actions.images.deleteImages]: (getState, dispatch, next, action) => {
+      // On image delete, also notify add-on and clean up localStorage
+      const images = action.payload;
+      postMessage("deleteImages", { images });
+      images.forEach(name => imageStorage.delete(name));
+      return next(action);
+    },
+    [actions.theme.clearCustomBackground]: (
+      getState,
+      dispatch,
+      next,
+      action
+    ) => {
+      // Watch for deleted backgrounds in current theme to purge orphaned images
+      const rv = next(action);
+      purgeImages(getState, dispatch);
+      return rv;
+    },
+    [actions.ui.setSavedThemes]: (getState, dispatch, next, action) => {
+      // Watch for changes in saved themes to purge orphaned images
+      const rv = next(action);
+      purgeImages(getState, dispatch);
+      return rv;
+    }
+  };
+
+  const purgeImages = (getState, dispatch) => {
```

Comment:
Basic garbage collection for unused images. Feels kind of dirty, but couldn't think of a better way to clean up since images are shared and a deletion in one theme doesn't mean it's not used in another.